core/amber/src/main/scala/edu/uci/ics/texera/web/service/ResultExportService.scala

```diff
@@ -362,4 +337,153 @@ class ResultExportService(opResultStorage: OpResultStorage, wId: UInteger) {
+      return ResultExportResponse("error", "No results to export")
+    }
+
+    val stream = new ByteArrayOutputStream()
+    val channel = new WritableByteChannelImpl(stream)
+    val allocator = new RootAllocator()
+
+    try {
+      val (writer, root) = createArrowWriter(results, allocator, channel)
+      try {
+        writeArrowData(writer, root, results)
+      } finally {
+        writer.close()
+        root.close()
+      }
+      finalizeArrowExport(request, user, stream, "arrow")
+    } finally {
+      allocator.close()
+      channel.close()
+    }
+  }
+
+  private def createArrowWriter(
+      results: Iterable[Tuple],
+      allocator: RootAllocator,
+      channel: WritableByteChannel
+  ): (ArrowFileWriter, VectorSchemaRoot) = {
+    val schema = results.head.getSchema
+    val arrowSchema = ArrowUtils.fromTexeraSchema(schema)
+    val root = VectorSchemaRoot.create(arrowSchema, allocator)
+    val writer = new ArrowFileWriter(root, null, channel)
+    (writer, root)
+  }
+
+  private def writeArrowData(
+      writer: ArrowFileWriter,
+      root: VectorSchemaRoot,
+      results: Iterable[Tuple]
+  ): Unit = {
+    writer.start()
+    val batchSize = 1000
+
+    // Convert to Seq to get total size
+    val resultSeq = results.toSeq
+    val totalSize = resultSeq.size
+
+    // Process in complete batches
+    for (batchStart <- 0 until totalSize by batchSize) {
+      val batchEnd = Math.min(batchStart + batchSize, totalSize)
+      val currentBatchSize = batchEnd - batchStart
+
+      // Process each tuple in the current batch
+      for (i <- 0 until currentBatchSize) {
+        val tuple = resultSeq(batchStart + i)
+        ArrowUtils.setTexeraTuple(tuple, i, root)
+      }
+
+      // Set the correct row count for this batch and write it
+      root.setRowCount(currentBatchSize)
+      writer.writeBatch()
+      root.clear()
+    }
+
+    writer.end()
+  }
+
+  private def finalizeArrowExport(
+      request: ResultExportRequest,
+      user: User,
+      stream: ByteArrayOutputStream,
+      extension: String
+  ): ResultExportResponse = {
+    val fileName = generateFileName(request, extension)
+    saveToDatasets(request, user, stream.toByteArray, fileName)
+    ResultExportResponse(
+      "success",
+      s"Arrow file saved as $fileName to Datasets ${request.datasetIds.mkString(",")}"
+    )
+  }
+
+  private def generateFileName(request: ResultExportRequest, extension: String): String = {
+    val latestVersion =
+      WorkflowVersionResource.getLatestVersion(UInteger.valueOf(request.workflowId))
+    val timestamp = LocalDateTime
+      .now()
+      .truncatedTo(ChronoUnit.SECONDS)
+      .format(DateTimeFormatter.ofPattern("yyyy-MM-dd_HH-mm-ss"))
+    sanitizePath(
+      s"${request.workflowName}-v$latestVersion-${request.operatorName}-$timestamp.$extension"
+    )
+  }
+
+  private def saveToDatasets(
+      request: ResultExportRequest,
+      user: User,
+      data: Array[Byte],
```

Comment:
here we are using an array of bytes, which means all the data are read into RAM. Can we use streams all the time? so that we don't maintain a copy of the data in RAM.